src/pdf_beamtime/src/pdf_beamtime_server.cpp

```diff
@@ -515,48 +547,45 @@ void PdfBeamtimeServer::handle_pause()
 
 void PdfBeamtimeServer::handle_stop()
 {
-  inner_state_machine_->set_internal_state(Internal_State::STOP);
+  RCLCPP_INFO(node_->get_logger(), "STOP command received");
+  inner_state_machine_->set_internal_state(Internal_State::CLEANUP);
+  execute_cleanup();
 }
 
-void PdfBeamtimeServer::execute_stop()
+void PdfBeamtimeServer::execute_cleanup()
 {
-  // @todo @ChandimaFernando This needs testing
+  RCLCPP_INFO(node_->get_logger(), "Cleanup is in progress");
+
   switch (current_state_) {
     case State::GRASP_SUCCESS:
-      // handle the decision to pick up something else after successfully grabing the sample
       inner_state_machine_->open_gripper();
-      set_current_state(State::RETRY_PICKUP);
-      inner_state_machine_->set_internal_state(Internal_State::RESTING);
-      RCLCPP_WARN(node_->get_logger(), "New goal needed !!!!");
       break;
 
     case State::PICKUP_RETREAT:
-      // Go back in states to pick up a different sample
-      set_current_state(State::GRASP_SUCCESS);
-      execute_stop();
-      break;
-
     case State::PLACE_APPROACH:
-      // Go back in states to pick up a different sample
-      set_current_state(State::PICKUP_RETREAT);
-      execute_stop();
-      break;
-
     case State::PLACE:
-      set_current_state(State::PICKUP_RETREAT);
-      inner_state_machine_->set_internal_state(Internal_State::RESTING);
-      RCLCPP_WARN(node_->get_logger(), "New goal needed !!!!");
+      return_sample();
```

Comment:
The way I read this is cleanup is
- HOME: no cleanup
- PICKUP_APPROACH: Do nothing. **Shouldn't I return home?**
- PICKUP: no cleanup?? **Shouldn't I back away from the sample and return home?**
- GRASP_SUCCESS: open gripper, stays in position around the sample. **Shouldn't I back away from the sample and return home?**
- GRASP_FAILURE: Do nothing. **Shouldn't I back away from the sample and return home?**
- PICKUP_RETREAT: returns sample
- PLACE_APPROACH: returns sample
- PLACE: returns sample
- RELEASE_SUCCESS: do nothing
- RELEASE_FAILURE: do nothing
- PLACE_RETREAT: do nothing

I see a couple of failure possibilities: 
1. If I clean up after a pause in pickup position, I'm in that position, but my state machine thinks I'm HOME. So then I send a new goal for a new sample which is 5 cm to the left. The robot navigates to the pickup_approach position diagonally, and knocks over the original sample. 
2. In any of the other cases, I pause and cleanup, then try to move the robot out of the beam. It collides with everything because it is not in the home position. 
3. In the return sample case, I end up in pickup_approach, which can lead to failure case 2.

Reply:
`reset_fsm` not only changes the state, but moves the robot home. Most of these can be ignored.

Reply:
PICKUP, GRASP_SUCCESS, GRASP_FAILURE now returns to PICKUP_APPROACH before returning home.
https://github.com/maffettone/erobs/blob/a934683ceb55e3e44db7b983139bd33b9a02334b/src/pdf_beamtime/src/pdf_beamtime_server.cpp#L568-L587